internal/cmd/commands/dev/dev.go

```diff
@@ -824,73 +823,84 @@ func (c *Command) Run(args []string) int {
 	c.opsServer = opsServer
 	c.opsServer.Start()
 
-	// Wait for shutdown
-	shutdownTriggered := false
-
-	// Add a force-shutdown goroutine to consume more interrupts
-	abortForceShutdownCh := make(chan struct{})
-	defer close(abortForceShutdownCh)
+	// On first interrupt, enter graceful shutdown for controllers and workers (workers wait for connections to drain)
+	// On second interrupt, force controller shutdown and send workers into shutdown (stop all connections)
+	// On third+ interrupt, force worker shutdown
+	shutdownDone := false
+	shutdownTriggerCount := uint32(0)
 
 	runShutdownLogic := func() {
-		go func() {
-			shutdownCh := make(chan os.Signal, 4)
-			signal.Notify(shutdownCh, os.Interrupt, syscall.SIGTERM)
-			for {
-				select {
-				case <-shutdownCh:
-					c.UI.Error("Forcing shutdown")
-					os.Exit(base.CommandUserError)
+		if sync.LoadUint32(&shutdownTriggerCount) == 1 {
+			go func() {
+				if c.Config.Controller != nil {
+					c.opsServer.WaitIfHealthExists(c.Config.Controller.GracefulShutdownWaitDuration, c.UI)
+				}
+
+				if !c.flagControllerOnly {
+					if !c.flagWorkerAuthStorageSkipCleanup && c.worker.WorkerAuthStorage != nil {
+						c.worker.WorkerAuthStorage.Cleanup()
+					}
+					if err := c.worker.GracefulShutdown(); err != nil {
+						c.UI.Error(fmt.Errorf("Error shutting down worker gracefully: %w", err).Error())
+					}
+					if err := c.worker.Shutdown(); err != nil {
```

Comment:
I'm not sure if this was possible before or not but we do run a risk here of calling `c.worker.Shutdown` more than once concurrently. That may not really be an issue, any errors that arise will be logged but probably be okay.

Reply:
The first check in worker.Shutdown() is whether or not shutdown has been initiated so we should be ok
```
	if !w.started.Load() {
		event.WriteSysEvent(w.baseContext, op, "already shut down, skipping")
		return nil
	}
```